enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md

```diff
@@ -0,0 +1,253 @@
+---
+title: cluster-logging-es-rollover-data-design
+authors:
+  - "@jcantril"
+reviewers:
+  - "@bparees"
+  - "@ewolinetz"
+  - "@richm"
+  - "@portante"
+  - "@pavolloffay"
+  - "@sichvoge"
+approvers:
+  - "@ewolinetz"
+  - "@richm"
+creation-date: 2019-11-04
+last-updated: 2019-11-04
+status: implementable
+---
+
+# Cluster Logging Elasticsearch Rollover Data Design
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] ~~Graduation criteria for dev preview, tech preview, GA~~
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+This proposal alters the data design for storing logs in Elasticsearch to co-locate logs
+to fewer indices.  It additionally leverages the [Elasticsearch rollover api](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/indices-rollover-index.html) to
+help maintain the number of indices and shards in order to align with Elastic's 
+[performance](https://www.elastic.co/guide/en/elasticsearch/guide/2.x/scale.html) 
+and scaling [recommendations](https://www.elastic.co/blog/how-many-shards-should-i-have-in-my-elasticsearch-cluster).
+
+## Motivation
+
+The initial data design for Cluster Logging segments logs by OpenShift namespace in order to facilitate multi-tenant support and data curation.  This choice was made because
+index level security was the only feature available from an open source library.  It additionally facilitated curation as full indicies could be removed when the retention period
+expired.  This means, however, at any one time, there are at least "**$noOfNamespaces * $daysRetatined**" number 
+of indices maintained by the Elasticsearch server.  Each index can additionally be sharded to spread the load across the Elasticsearch nodes.  The end
+result of this shard explosion is the Elasticsearch cluster performance is not optimized and is not capable of efficiently processing logs or storing as many as it could. 
+
+Each index adds load and overhead (e.g. mapping, metadata) to the Elasticsearch cluster that needs to be tracked beyond the actual data.   Elasticsearch has recommendations for maximum shard size and the number of shards per node per allocated gigabyte of heap.  Cluster logging typically exceeds these recommendations for any OpenShift clusters that has significant log traffic.
```

Comment:
do we exceed it in shard size, or shard count?  reducing the number of indexes would, i assume, increase the size of the shards, right?

Reply:
Yes, which is where Elasticsearch is most efficient and its design point.

Reply:
ok.  the statements in this EP says "ES has recommendations for maximum shard size + number of shards per node.  cluster logging typically exceeds these recommendations..."

So if we're already exceeding the shard size recommendation, doing something that would make the shards bigger would seem problematic.  So is the statement incorrect?  Are we *not* currently exceeding shard size recommendations?  Or is there something else i'm missing here?

Reply:
The new rollover scheme prevents us from exceeding the size recommendations, and keeps the number of shards down by leveraging Document Level Security to store all logs from different namespaces in the same set of indices.

So today, we have 3 shards per index (3 ES pods), one index per namespace per day.  So for 7 day retention of 100 namespaces, we have 3 * 7 * 100, or 2,100 shards.  We don't put an upper bound on data per shard, so heavy logging traffic can grow shards above the recommended size.

With the new scheme, with 7 day retention, if all the data in a day is less than the recommended size, we'll have 3 * 7 shards, 21 shards.  If the data gathered in a day is 3 times the recommended size, then we'll have 63 shards.  If it is 10 times the recommended size, we'll have 210 shards.  All significantly less than the current scheme and must more efficient because the HEAP used to track all the shards is reduced.

Reply:
ok, i missed the part where we are going to cap the shard size and introduce new shards when the size is reached.

So there is no risk of us exceeding recommended shard *size*.

We may still exceed recommended shard count, depending on configured retention limits and data logging rates.

Reply:
Possibly that is why we need https://github.com/containers/conmon/issues/84 accepted in to `conmon` and then by CRI so that Kubernetes can be plummed to give somebody a hope of managing byte-capture of `stdout`/`stderr`.